### PR TITLE
Fix storage pruning with rollback

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: process/block/metablock.go [hunk 1]
```diff
@@ -982,6 +982,10 @@ func (mp *metaProcessor) CommitBlock(
 
 	chainHandler.SetCurrentBlockHeaderHash(headerHash)
 
+	if lastMetaBlock != nil && lastMetaBlock.IsStartOfEpochBlock() {
```
Comment (sasurobert) [inline process/block/metablock.go]: check.ifNil as lastMetaBock comes from chainHandler and it is of type data.HeaderHandler

File: process/block/shardblock.go [hunk 2]
```diff
@@ -377,7 +377,7 @@ func (sp *shardProcessor) checkEpochCorrectness(
 
 	isOldEpochAndShouldBeNew := sp.epochStartTrigger.IsEpochStart() &&
 		header.GetRound() > sp.epochStartTrigger.EpochFinalityAttestingRound()+process.EpochChangeGracePeriod &&
-		header.GetEpoch() < sp.epochStartTrigger.Epoch()
+		header.GetEpoch() < sp.epochStartTrigger.Epoch() && sp.epochStartTrigger.EpochStartRound() < sp.epochStartTrigger.EpochFinalityAttestingRound()
```
Comment (sasurobert) [inline process/block/shardblock.go]: move last condition to new line.

File: process/peer/process.go [hunk 3]
```diff
@@ -334,9 +334,6 @@ func (vs *validatorStatistics) UpdatePeerState(header data.HeaderHandler) ([]byt
 }
 
 func (vs *validatorStatistics) displayRatings() {
```
Comment (sasurobert) [inline process/peer/process.go]: delete function as well.

File: process/peer/process.go [hunk 4]
```diff
@@ -730,6 +733,32 @@ func (vs *validatorStatistics) getTempRating(s string) uint32 {
 	return peer.GetTempRating()
 }
 
+func (vs *validatorStatistics) display(s string) {
```
Comment (sasurobert) [inline process/peer/process.go]: maybe change s to blsPubKey, or accountKey

File: storage/pruning/pruningStorer.go [hunk 5]
```diff
@@ -528,13 +540,43 @@ func (ps *PruningStorer) changeEpoch(epoch uint32) error {
...
+
+	oldestEpochActive := int64(epoch) - int64(ps.numOfActivePersisters) + 1
+	if oldestEpochActive < 0 {
+		oldestEpochActive = 0
+	}
+
+	for e := int64(epoch); e >= oldestEpochActive; e-- {
+		p, ok := ps.persistersMapByEpoch[uint32(e)]
+		if !ok {
+			return nil
```
Comment (sasurobert) [inline storage/pruning/pruningStorer.go]: why not continue ?

Comment (sasurobert) [reply storage/pruning/pruningStorer.go]: is this a critical error ?

File: process/block/shardblock.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/block/shardblock.go]: This will probably never happen in shard

File: process/block/metablock.go (hunk 1 above)
Comment (AdoAdoAdo) [reply process/block/metablock.go]: done

File: process/block/shardblock.go (hunk 2 above)
Comment (AdoAdoAdo) [reply process/block/shardblock.go]: done

File: process/peer/process.go (hunk 3 above)
Comment (AdoAdoAdo) [reply process/peer/process.go]: done

File: process/peer/process.go (hunk 4 above)
Comment (AdoAdoAdo) [reply process/peer/process.go]: changed to validatorKey

**Pull request merged**